syncer/find_fork.go

```diff
@@ -0,0 +1,318 @@
+	"github.com/spacemeshos/go-spacemesh/common/util"
+	"github.com/spacemeshos/go-spacemesh/fetch"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/layers"
+)
+
+var ErrPeerMeshChangedMidSession = errors.New("peer mesh changed mid session")
+
+type layerHash struct {
+	lid     types.LayerID
+	hash    types.Hash32
+	created time.Time
+}
+
+func newLayerHash(lid types.LayerID, hash types.Hash32) *layerHash {
+	return &layerHash{
+		lid:  lid,
+		hash: hash,
+	}
+}
+
+// boundary is used to define the to and from layers in the hash mesh requests to peers.
+// the hashes of the boundary layers are known to the node and are used to double-check that
+// the peer has not changed its opinions on those layers.
+// if the boundary hashes change during the fork-finding session, the session is aborted.
+type boundary [2]*layerHash
+
+func newBoundary() *boundary {
+	return &boundary{
+		newLayerHash(types.LayerID{}, types.Hash32{}),
+		newLayerHash(types.LayerID{}, types.Hash32{}),
+	}
+}
+
+func (b *boundary) setFrom(lh *layerHash) {
+	b[0] = lh
+}
+
+func (b *boundary) setTo(lh *layerHash) {
+	b[1] = lh
+}
+
+func (b *boundary) fromLayer() types.LayerID {
+	return b[0].lid
+}
+
+func (b *boundary) fomHash() types.Hash32 {
+	return b[0].hash
+}
+
+func (b *boundary) toLayer() types.LayerID {
+	return b[1].lid
+}
+
+func (b *boundary) toHash() types.Hash32 {
+	return b[1].hash
+}
+
+func (b *boundary) MarshalLogObject(encoder log.ObjectEncoder) error {
+	encoder.AddUint32("from", b.fromLayer().Uint32())
+	encoder.AddString("from_hash", b.fomHash().ShortString())
+	encoder.AddUint32("to", b.toLayer().Uint32())
+	encoder.AddString("to_hash", b.toHash().ShortString())
+	return nil
+}
+
+type ForkFinder struct {
+	logger           log.Log
+	db               sql.Executor
+	fetcher          fetcher
+	maxHashesInReq   uint32
+	maxStaleDuration time.Duration
+
+	mu          sync.Mutex
+	agreedPeers map[p2p.Peer]*layerHash
+}
+
+func NewForkFinder(lg log.Log, db sql.Executor, f fetcher, maxHashes uint32, maxStale time.Duration) *ForkFinder {
+	return &ForkFinder{
+		logger:           lg,
+		db:               db,
+		fetcher:          f,
+		maxHashesInReq:   maxHashes,
+		maxStaleDuration: maxStale,
+		agreedPeers:      make(map[p2p.Peer]*layerHash),
+	}
+}
+
+// Purge cached agreements with peers.
+func (ff *ForkFinder) Purge(all bool, toPurge ...p2p.Peer) {
+	ff.mu.Lock()
+	defer ff.mu.Unlock()
+
+	if all {
+		ff.agreedPeers = make(map[p2p.Peer]*layerHash)
+		return
+	}
+
+	if len(toPurge) > 0 {
+		for _, p := range toPurge {
+			delete(ff.agreedPeers, p)
+		}
+		return
+	}
+
+	uniquePeers := util.SliceToSetStringer(ff.fetcher.GetPeers())
+	for p, lh := range ff.agreedPeers {
+		if _, ok := uniquePeers[p.String()]; !ok {
+			if time.Since(lh.created) >= ff.maxStaleDuration {
+				delete(ff.agreedPeers, p)
+			}
+		}
+	}
```

Comment:
Can `ff.fetcher.GetPeers()` return a slice of peers that contains duplicates? If not this would work as well:

```suggestion
	for _, p := range ff.fetcher.GetPeers() {
		if lh, ok := ff.agreedPeers[p.String()]; !ok {
			if time.Since(lh.created) >= ff.maxStaleDuration {
				delete(ff.agreedPeers, p)
			}
		}
	}
```

This could also be faster, depending on how big we expect the slice returned by `ff.fetch.GetPeers()` and the map `ff.agreedPeers` to be.

Reply:
Even if it returns duplicates, it's still at least as fast as first converting it to a set and then iterating over the set would be.

Reply:
it is a little different. here i want to remove items in `agreedPeers` that's not in `fetcher.GetPeers()`.
i ended up created a temp map here. (also part of Dmitry's feedback)